### DISCUSSION TITLE
Multi-select suggester

### DISCUSSION THREAD

**Discussion by Zachatoo** [SilentVoid13/Templater#1099] [Templates Showcase]
We can (kindof) do a multi-select suggester by sticking the suggester in a while loop, letting the user keep selecting options and adding them to a list of selected options until the user hits the escape key to exit out of the loop.

User script (docs [here](https://silentvoid13.github.io/Templater/user-functions/script-user-functions.html)):
```js
/**
 * Spawns a multi-select suggester prompt and returns the user's chosen items.
 * @param {object} tp Templater tp object.
 * @param {string[] | ((item: T) => string)} textItems Array of strings representing the text that will be displayed for each item in the suggester prompt. This can also be a function that maps an item to its text representation.
 * @param {T[]} items Array containing the values of each item in the correct order.
 * @param {boolean} throwOnCancel Throws an error if the prompt is canceled, instead of returning a null value.
 * @param {string} placeholder Placeholder string of the prompt.
 * @param {number} limit Limit the number of items rendered at once (useful to improve performance when displaying large lists).
 * @returns A list of selected items.
 */
async function multiSuggester(
  tp,
  textItems,
  items,
  throwOnCancel = false,
  placeholder = "",
  limit = undefined
) {
  // List of items that are selected in the suggester
  const selectedItems = [];
  // Looping to keep suggester modal open until escape is hit
  while (true) {
    const selectedItem = await tp.system.suggester(
      textItems,
      items,
      throwOnCancel,
      placeholder,
      limit
    );
    // If escape is hit, break out of loop to close suggester modal
    if (!selectedItem) {
      break;
    }
    // Otherwise, add selected item to list of selected items, remove item from multi-select, and keep looping
    selectedItems.push(selectedItem);
    items = items.filter((item) => item !== selectedItem);
  }
 
  return selectedItems;
}
 
module.exports = multiSuggester;
```

Sample usage in template:
```js
<% tp.user.multiSuggester(tp, item => item, ["Value1", "Value2"]) %>
```

Originally posted [here](https://zachyoung.dev/snippets/templater-multi-select-suggester).

**Comment by AB1908** [comment]
How would you do a true multi select? Just use a series of checkboxes? I've been considering adding one given that I've been looking into doing some fixes here and there but I haven't needed one personally so it's fallen down the pecking order haha.

**Reply by Zachatoo** [reply]
I do prefer multi select designs that have checkboxes. At the very least having checkmarks next to selected options to make it more obvious what options you've selected.

There's some other features to consider as well

- "Select all" option
- Show all selected options in or above the input box
- Ability to have some options pre-selected (pass in all options and selected options)
- Ensure that checkboxes match user's theme (use Obsidian's CSS variables)

Definitely would be cool for a feature like this to be built in, rather than user's having to search the internet to find alternative solutions. Though I haven't seen this requested much, happened to see two requests for it in the same day after not seeing it for many months so I made this example.
